src/hashtable.c

```diff
@@ -0,0 +1,2135 @@
+    if (hashtableSize(ht) == 0) return NULL;
+    uint64_t hash = hashKey(ht, key);
+    int pos_in_bucket = 0;
+    int table_index = 0;
+    bucket *b = findBucket(ht, hash, key, &pos_in_bucket, &table_index);
+    if (b) {
+        hashtablePauseRehashing(ht);
+
+        /* Store position. */
+        assert(p != NULL);
+        p->bucket = b;
+        p->pos_in_bucket = pos_in_bucket;
+        p->table_index = table_index;
+        return &b->entries[pos_in_bucket];
+    } else {
+        return NULL;
+    }
+}
+
+/* Clears the position of the entry in the hashtable and resumes rehashing. The
+ * entry destructor is NOT called. The position is acquired using a preceding
+ * call to hashtableTwoPhasePopFindRef(). */
+void hashtableTwoPhasePopDelete(hashtable *ht, hashtablePosition *position) {
+    /* Read position. */
+    struct position *p = positionFromOpaque(position);
+    bucket *b = p->bucket;
+    int pos_in_bucket = p->pos_in_bucket;
+    int table_index = p->table_index;
+
+    /* Delete the entry and resume rehashing. */
+    assert(isPositionFilled(b, pos_in_bucket));
+    b->presence &= ~(1 << pos_in_bucket);
+    ht->used[table_index]--;
+    hashtableShrinkIfNeeded(ht);
+    hashtableResumeRehashing(ht);
+    if (b->chained && !hashtableIsRehashingPaused(ht)) {
+        /* Rehashing paused also means bucket chain compaction paused. It is
+         * paused while iterating and when a scan callback is running, to be
+         * able to live up to the scan and iterator guarantees. In those cases,
+         * we do the compaction in the scan and iterator code instead. */
+        fillBucketHole(ht, b, pos_in_bucket, table_index);
+    }
+}
+
+/* Initializes the state for an incremental find operation.
+ *
+ * Incremental find can be used to speed up the loading of multiple objects by
+ * utilizing CPU branch predictions to parallelize memory accesses. Initialize
+ * the data for a number of incremental find operations. Then call
+ * hashtableIncrementalFindStep on them in a round-robin order until all of them
+ * are complete. Finally, if necessary, call hashtableIncrementalFindGetResult.
+ */
+void hashtableIncrementalFindInit(hashtableIncrementalFindState *state, hashtable *ht, const void *key) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    if (hashtableSize(ht) == 0) {
+        data->state = HASHTABLE_NOT_FOUND;
+    } else {
+        data->state = HASHTABLE_NEXT_BUCKET;
+        data->bucket = NULL;
+        data->hashtable = ht;
+        data->key = key;
+        data->hash = hashKey(ht, key);
+    }
+}
+
+/* Returns 1 if more work is needed, 0 when the done. Call this function
+ * repeatedly until it returns 0. Then use hashtableIncrementalFindGetResult to
+ * fetch the result. */
+int hashtableIncrementalFindStep(hashtableIncrementalFindState *state) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    switch (data->state) {
+    case HASHTABLE_CHECK_ENTRY:
```

Comment:
I wonder if we can also keep a nextBucket pointer in the state and prefetch the next bucket (incase the current one is chained) at this point?

Reply:
In `HASHTABLE_NEXT_ENTRY`, where we have a bucket and we're looking for an entry with a matching has to prefetch, there we could prefetch all the entries with matching hashes and also the next bucket, but it's unlikely that an entry with a matching hash is not a match, so I skipped this for simplicity. If there is no entry with a matching hash, we prefetch the next bucket immediately.

If you have an implementation that's not too complex, please share the code.

Reply:
right, sorry I meant HASHTABLE_NEXT_ENTRY. I agree with you this might be a low chance. we can test for it later on